File: javascript/src/rules.ts
import { JsonLogicRule } from "./extend-JsonLogic"


export interface Rule {
    id: string
    active: boolean
    businessDescription?: string
    description: string
    logic: JsonLogicRule
}
export type Ruleset = Rule[]
export const euTemplateRuleset: Ruleset = require("../../../dgc-business-rules-testdata/EU")


export const valueSets = require("../../../dgc-business-rules-testdata/valuesets/valueSets.json")


export type RuleRunner = (rule: Rule, payload: any, validationClock?: string) => any

export const ruleRunner = <T>(evaluate: (expr: T, data: any) => any): RuleRunner => (rule: Rule, payload: any, validationClock?: string) => {
    const data = {
        external: {
            valueSets,
            validationClock,
            "countryCode": "NL",
            "exp": "2022-10-21T18:25:43-05:00"
        },
        payload
    }
    return evaluate(rule.logic as unknown as T, data)
}

File: rules-runner/rules-runner-kotlin/src/main/kotlin/eu/ehn/dcc/rulesets/Rules.kt
package eu.ehn.dcc.rulesets

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.node.JsonNodeFactory
import com.fasterxml.jackson.databind.node.ObjectNode
import eu.ehn.dcc.certlogic.evaluate
import java.io.File


data class Rule(
    var id: String,
    var active: Boolean,
    var businessDescription: String?,
    var description: String,
    var inputParameter: String?,
    var certLogicExpression: JsonNode
)
typealias Ruleset = List<Rule>

val rulesPath = File("../../rulesets")
val euTemplateRuleset: Ruleset = readJson(rulesPath / "EU" / "template-ruleset.json")

val valueSets: ObjectNode = readJson(rulesPath / "valueSets.json")


fun Rule.run(payload: JsonNode, validationClock: String?): JsonNode {
    val data = JsonNodeFactory.instance.objectNode()
        .set<ObjectNode>(
            "external",
            JsonNodeFactory.instance.objectNode()
                .set<ObjectNode>("valueSets", valueSets)
                .put("validationClock", validationClock)
                .put("countryCode", "CZ")
                .put("exp", "2022-10-21T18:25:43-05:00")
        )
        .set<ObjectNode>("payload", payload)
    return evaluate(this.certLogicExpression, data)
}

